File: RxTango/python/examples/retry.py
"""Retry on transient errors using ops.retry.

Demonstrates ops.retry(n) for simple retries and exponential-backoff retry
using rx.timer + ops.catch for smarter recovery.

Mirrors `TangoTestRetry.java`.

Usage:
    python retry.py [device] [interval-ms]

    device       defaults to tango://localhost:10000/sys/tg_test/1
    interval-ms  poll interval; defaults to 500
"""

import asyncio
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute


def make_backoff_retry(max_retries: int = 5, base_delay_ms: int = 500):
    """Return an operator that retries with exponential backoff."""
    attempt = [0]

    def handler(source: rx.Observable) -> rx.Observable:
        def catch_fn(exc, _src):
            attempt[0] += 1
            if attempt[0] > max_retries:
                return rx.throw(exc)
            delay = timedelta(milliseconds=base_delay_ms * (2 ** (attempt[0] - 1)))
            print(f"  ERROR: {exc}  (retry {attempt[0]}/{max_retries} in {delay.total_seconds():.1f}s)")
            return rx.timer(delay).pipe(
                ops.flat_map(lambda _: source)
            )

        return source.pipe(ops.catch(catch_fn))

    return lambda src: rx.defer(lambda: make_backoff_retry(max_retries, base_delay_ms)(src))


async def main() -> None:
    device      = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    interval_ms = int(sys.argv[2]) if len(sys.argv) > 2 else 500

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    print(f"Polling with retry on {device}  (Ctrl+C to stop)\n")
    print(f"  {'value':>20}")
    print("  " + "-" * 22)

    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        # Simple retry: up to 3 immediate retries per tick on error
        ops.flat_map(
            lambda _: read_attribute(device, "double_scalar").pipe(
                ops.retry(3)
            )
        ),
    ).subscribe(
        on_next=lambda v: print(f"  {v:>+20.6f}"),
        on_error=lambda e: print(f"  FATAL: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n  stopped.")

File: RxTango/cpp/examples/retry.cpp
/**
 * Retry patterns — how to handle transient failures in a polling pipeline.
 *
 * Three patterns:
 *   fixed   — retry(N) on the inner read; whole pipeline retries N times
 *   backoff — exponential backoff via zip+delay (sketch)
 *   inner   — retry inside flat_map; outer pipeline never sees the error
 *
 * Mirrors Python's retry.py and Java's TangoTestRetry.java.
 *
 * Usage:
 *   ./retry [strategy: fixed|inner] [device] [attr] [retries] [interval-ms]
 *   defaults: inner  sys/tg_test/1  double_scalar  3  500
 */

#include <chrono>
#include <csignal>
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string strategy   = argc > 1 ? argv[1] : "inner";
    const std::string device     = argc > 2 ? argv[2] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr       = argc > 3 ? argv[3] : "double_scalar";
    const int         retries    = argc > 4 ? std::stoi(argv[4]) : 3;
    const int         interval_ms= argc > 5 ? std::stoi(argv[5]) : 500;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Retry strategy: " << strategy
              << "  retries=" << retries << "  (Ctrl+C to stop)\n\n";

    if (strategy == "fixed") {
        // retry(N) on the outer polling chain
        rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
            .flat_map([device, attr](long) {
                return rxtango::read_attribute<double>(device, attr);
            })
            .retry(retries)
            .subscribe(
                [](double v) { std::cout << "  value: " << v << "\n"; },
                [](std::exception_ptr e) {
                    try { std::rethrow_exception(e); }
                    catch (std::exception& ex) {
                        std::cerr << "  FINAL ERROR after retries: " << ex.what() << "\n";
                    }
                }
            );
    } else {
        // inner retry — each read retries independently; outer pipeline sees only successes
        rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
            .flat_map([device, attr, retries](long) {
                return rxtango::read_attribute<double>(device, attr)
                    .retry(retries)
                    .on_error_resume_next([](std::exception_ptr) {
                        // exhausted retries → emit NaN so the pipeline continues
                        return rxcpp::observable<>::just(
                            std::numeric_limits<double>::quiet_NaN());
                    });
            })
            .filter([](double v) { return !std::isnan(v); })
            .subscribe(
                [](double v) { std::cout << "  value: " << v << "\n"; },
                [](std::exception_ptr e) {
                    try { std::rethrow_exception(e); }
                    catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
                }
            );
    }

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    return 0;
}
